tests/regression/test_fdm.py

```diff
@@ -0,0 +1,203 @@
+        else:
+            family = "DQ" if element == "dg" else "Q"
+        return VectorFunctionSpace(mesh, family, degree)
+
+
+@pytest.mark.skipcomplex
+def test_direct_solver(fs):
+    mesh = fs.mesh()
+    x = SpatialCoordinate(mesh)
+    u_exact = dot(x, x)
+    ncomp = fs.ufl_element().value_size()
+    if ncomp > 1:
+        u_exact = as_vector([u_exact + Constant(k) for k in range(ncomp)])
+
+    n = FacetNormal(mesh)
+    f_exact = grad(u_exact)
+    B = u_exact - div(f_exact)
+    T = dot(f_exact, n)
+    uh = Function(fs)
+    u = TrialFunction(fs)
+    v = TestFunction(fs)
+
+    subs = (1, 3)
+    if mesh.layers:
+        subs += ("top",)
+
+    bcs = [DirichletBC(fs, u_exact, sub) for sub in subs]
+
+    # sub_Dir = vertical subdomains for Dirichlet BCs
+    # sub_Neu = vertical subdomains for Neumann BCs
+    sub_Dir = "everywhere" if "on_boundary" in subs else tuple(s for s in subs if type(s) == int)
+    if sub_Dir == "everywhere":
+        sub_Neu = ()
+    else:
+        sub_Neu = tuple(set(mesh.exterior_facets.unique_markers) - set(s for s in subs if type(s) == int))
+
+    if mesh.layers:
+        dS_int = dS_v + dS_h
+        ds_Dir = ds_v(sub_Dir)
+        ds_Neu = ds_v(sub_Neu)
+        if "bottom" in subs:
+            ds_Dir += ds_b
+        else:
+            ds_Neu += ds_b
+        if "top" in subs:
+            ds_Dir += ds_t
+        else:
+            ds_Neu += ds_t
+    else:
+        dS_int = dS
+        ds_Dir = ds(sub_Dir)
+        ds_Neu = ds(sub_Neu)
+
+    N = fs.ufl_element().degree()
+    try:
+        N, = set(N)
+    except TypeError:
+        pass
+
+    eta = Constant((N+1)**2)
+    h = CellVolume(mesh)/FacetArea(mesh)
+    penalty = eta/h
+
+    a = (inner(v, u)*dx
+         + inner(grad(v), grad(u))*dx
+         + inner(outer_jump(v, n), avg(penalty) * outer_jump(u, n)) * dS_int
+         - inner(avg(grad(v)), outer_jump(u, n)) * dS_int
+         - inner(avg(grad(u)), outer_jump(v, n)) * dS_int
+         + inner(v, penalty * u) * ds_Dir
+         - inner(grad(v), outer(u, n)) * ds_Dir
+         - inner(grad(u), outer(v, n)) * ds_Dir)
+
+    L = (inner(v, B)*dx
+         + inner(v, T)*ds_Neu
+         + inner(v, penalty * u_exact) * ds_Dir
+         - inner(grad(v), outer(u_exact, n)) * ds_Dir)
+
+    problem = LinearVariationalProblem(a, L, uh, bcs=bcs)
+    solver = LinearVariationalSolver(problem, solver_parameters={
+        "mat_type": "matfree",
+        "ksp_type": "cg",
+        "ksp_atol": 0.0E0,
+        "ksp_rtol": 1.0E-8,
+        "ksp_max_it": 20,
+        "ksp_monitor": None,
+        "ksp_norm_type": "unpreconditioned",
+        "pc_type": "python",
+        "pc_python_type": "firedrake.FDMPC",
+        "fdm_pc_type": "cholesky",
+        "fdm_pc_factor_mat_solver_type": "mumps",
+        "fdm_pc_factor_mat_ordering_type": "nd",
+    }, appctx={"eta": eta, })
+    solver.solve()
+
+    assert solver.snes.ksp.getIterationNumber() == 1 and norm(u_exact-uh, "H1") < 1.0E-8
```

Comment:
The tests don't appear to exercise the full range of functionality. For example,  could there be a test that uses spatially  varying coefficients,  as well as lower order terms?